src/Orleans/Core/GrainFactory.cs

```diff
@@ -85,6 +97,182 @@ public static TGrainInterface GetGrain<TGrainInterface>(string primaryKey, strin
+            where TGrainObserverInterface : IGrainObserver
+        {
+            var interfaceType = typeof(TGrainObserverInterface);
+            if (!interfaceType.IsInterface)
+            {
+                throw new ArgumentException(
+                    string.Format(
+                        "The provided type parameter must be an interface. '{0}' is not an interface.",
+                        interfaceType.FullName));
+            }
+
+            if (!interfaceType.IsInstanceOfType(obj))
+            {
+                throw new ArgumentException(
+                    string.Format("The provided object must implement '{0}'.", interfaceType.FullName),
+                    "obj");
+            }
+            
+            Delegate creator;
+
+            if (!referenceCreators.TryGetValue(interfaceType, out creator))
+            {
+                creator = referenceCreators.GetOrAdd(interfaceType, CreateCreateObjectReferenceDelegate);
+            }
+
+            var resultTask = ((Func<TGrainObserverInterface, Task<TGrainObserverInterface>>)creator)((TGrainObserverInterface)obj);
+            return resultTask;
+        }
+
+        /// <summary>
+        /// Deletes the provided object reference.
+        /// </summary>
+        /// <typeparam name="TGrainObserverInterface">
+        /// The specific <see cref="IGrainObserver"/> type of <paramref name="obj"/>.
+        /// </typeparam>
+        /// <param name="obj">The reference being deleted.</param>
+        /// <returns>A <see cref="Task"/> representing the work performed.</returns>
+        public static Task DeleteObjectReference<TGrainObserverInterface>(
+            IGrainObserver obj) where TGrainObserverInterface : IGrainObserver
+        {
+            var interfaceType = typeof(TGrainObserverInterface);
+            if (!interfaceType.IsInterface)
+            {
+                throw new ArgumentException(
+                    string.Format(
+                        "The provided type parameter must be an interface. '{0}' is not an interface.",
+                        interfaceType.FullName));
+            }
+
+            if (!interfaceType.IsInstanceOfType(obj))
+            {
+                throw new ArgumentException(
+                    string.Format("The provided object must implement '{0}'.", interfaceType.FullName),
+                    "obj");
+            }
+
+            Delegate destroyer;
+
+            if (!referenceDestoyers.TryGetValue(interfaceType, out destroyer))
+            {
+                destroyer = referenceDestoyers.GetOrAdd(interfaceType, CreateDeleteObjectReferenceDelegate);
+            }
+
+            return ((Func<TGrainObserverInterface, Task>)destroyer)((TGrainObserverInterface)obj);
+        }
+
+        #region IGrainObserver Methods
+
+
+        private static Delegate CreateCreateObjectReferenceDelegate(Type interfaceType)
+        {
+            var grainFactoryName = TypeUtils.GetSimpleTypeName(interfaceType, t => false);
```

Comment:
Sorry for delay, I got interrupted twice (for several hours) trying to finish this code review.

Can we factor out this duplicated reflection related code between MakeCreate*, MakeDelete* and CreateCaster into a shared helper method? I think we need a separate issue (I'll open it after this one gets in) to completely remove this dependency on generated code. Otherwise, we can't really get rid of the factory generation.